### ISSUE TITLE
Key Generation during login is a feature not a bug.

### ISSUE THREAD

**Issue by CANAWEN** [CryptoverseCC/ethmail.cc#23]
From the FAQ:
> For logins this would be simple. Just generate key on first login, encrypt it with password and store it on ETHMail server in encrypted form (much like how ProtonMail does it). That’s ok and probably the best for initial implementation, but what about email coming to your mailbox even before you logged in for the first time?

Email coming into your mailbox before account creation should be treated like original email treats messages sent to accounts that aren’t created yet, it trips an error and the sender gets a message saying “no address exists / is invalid”.

I see nothing wrong with this. In fact, imagine the spam waiting for if you senders could send emails to any email account before you create it, it’d be madness.

**Comment by gkapkowski** [comment]
That's actually against ETHMail most important principle. There is no account creation. All Ethereum addresses have associated mailbox and anyone can send an email to any address even before the owner of the account has any knowledge of ETHMail. 

The spam problem is a different one. It's a matte of filtering and marking messages appropriately.

**Comment by iceguru** [comment]
The downside ofc is that these emails still follow the 45 day term to my knowledge. So I'll never know if I had emails I missed.

**Comment by gkapkowski** [comment]
Yes, I might actually change that, I would probably be able to change the 45 days into deleting largest messages (keeping some grace period before actual removal) when the inbox space reaches above 5-10MB for free accounts. This should be plenty of space for normal communication to sit and wait for the recipient while not costing me too much. What do you think?

**Comment by ToolsDevler** [comment]
I'm doing research in this and try out technical concepts since some months and I, of course, hit my toe at the same issue.

Of course the security implications make this really hard, which is why I'm also still working on it.

I try to work this around with a smart contract. There you can register a mapping between your wallet address and the public key into a mapping and exchange public keys. I even think about creating a new wallet for every set of parties in a conversation (e.g. between Bob and Alice and between Bob James and Alice for multiple receivers or mailing lists) to improve security and ease security related measures. This is, because the wallet used for communication only is used by the conversation attendees and this doesn't touch your main wallet (where you might store money). The management of those related wallets might also make the public/private key management a lot easier.

So the smart contract stores (keyed by msg.sender to prevent spying/tempering) all the mappings between the own wallet and the messaging (or exchange) wallets. Also the mails itself can be stored in the smart contract or at least the block height and transaction id. This should also make it possible to catch prior mails without parsing all transactions first.

I know there are still unanswered questions, but for me this looks like a promising approach.

In parallel I'm checking out the possibilities with IPFS, which is where I discovered [ipmail](https://github.com/Geo25rey/ipmail).

What do you think?
